### PR TITLE
Add Syscalls: Neo.Storage.GetReadOnlyContext and Neo.StorageContext.AsReadOnly

### LINKED ISSUE #177: Add Neo.Storage.GetReadOnlyContext and Neo.StorageContext.AsReadOnly syscalls

### PR DESCRIPTION
**Pull request by davemneo**
resolves #177

Update imports:
  ArgumentsHelper.ts
  compile/syscalls.ts
  generateTypes.ts

compile/syscalls.ts
 - TypeAlias update == to =
 - add support for Neo.Storage.GetReadOnlyContext and Neo.StorageContext.AsReadOnly

package.json
 - update generate-sc-types

### CODE REVIEW

**Comment by davemneo** [comment]
Should there be an entry in 
https://github.com/davemneo/neo-one/blob/4b237c274e251574d9667334396511a4e437422f/packages/neo-one-node-vm/src/syscalls.ts#L1583

for Neo.StorageContext.AsReadOnly

File: packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts [hunk 1]
```diff
@@ -977,6 +977,15 @@ export const SYSCALLS = {
     name: 'Neo.Storage.GetContext',
     returnType: StorageContextValue,
   }),
+  'Neo.Storage.GetReadOnlyContext': new SimpleSysCall({
+    name: 'Neo.Storage.GetReadOnlyContext',
+    returnType: StorageContextValue,
```
Comment (dicarlo2) [inline packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts]: Actually, let's have this return a `StorageContextReadonlyValue`.

File: packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts [hunk 2]
```diff
@@ -977,6 +977,15 @@ export const SYSCALLS = {
...
+  }),
+  'Neo.StorageContext.AsReadOnly': new SimpleSysCall({
+    name: 'Neo.StorageContext.AsReadOnly',
+    args: [new SysCallArgument('context', StorageContextValue)],
+    returnType: StorageContextValue,
```
Comment (dicarlo2) [inline packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts]: Same here

File: packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts [hunk 3]
```diff
@@ -977,6 +977,15 @@ export const SYSCALLS = {
...
+  }),
   'Neo.Storage.Get': new SimpleSysCall({
     name: 'Neo.Storage.Get',
     args: [new SysCallArgument('context', StorageContextValue), new SysCallArgument('key', StorageKey)],
```
Comment (dicarlo2) [inline packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts]: And then have this take in a union of `StorageContextValue` and `StorageContextReadonlyValue`

Comment (dicarlo2) [reply packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts]: Same for elsewhere that are read operations ('Neo.Storage.Find' might be the only other place)

File: packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts [hunk 4]
```diff
@@ -985,7 +995,7 @@ export const SYSCALLS = {
   'Neo.Storage.Find': new SimpleSysCall({
     name: 'Neo.Storage.Find',
     args: [new SysCallArgument('context', StorageContextValue), new SysCallArgument('prefix', StorageKey)],
-    returnType: StorageIteratorValue,
+    returnType: StorageContextReadonlyValue,
```
Comment (dicarlo2) [inline packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts]: Why was this changed?

File: packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts [hunk 5]
```diff
@@ -977,6 +978,15 @@ export const SYSCALLS = {
     name: 'Neo.Storage.GetContext',
     returnType: StorageContextValue,
   }),
+  'Neo.Storage.GetReadOnlyContext': new SimpleSysCall({
+    name: 'Neo.Storage.GetReadOnlyContext',
+    returnType: StorageContextReadonlyValue,
+  }),
+  'Neo.StorageContext.AsReadOnly': new SimpleSysCall({
+    name: 'Neo.StorageContext.AsReadOnly',
+    args: [new SysCallArgument('context', new UnionValue([StorageContextValue, StorageContextReadonlyValue]))],
```
Comment (dicarlo2) [inline packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts]: This should take a `StorageContextValue`

File: packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts [hunk 6]
```diff
@@ -985,7 +995,7 @@ export const SYSCALLS = {
   'Neo.Storage.Find': new SimpleSysCall({
     name: 'Neo.Storage.Find',
     args: [new SysCallArgument('context', StorageContextValue), new SysCallArgument('prefix', StorageKey)],
```
Comment (dicarlo2) [inline packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts]: This should take a union of `StorageContextValue` and `StorageContextReadonlyValue`

File: packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts [hunk 7]
```diff
@@ -977,6 +978,15 @@ export const SYSCALLS = {
...
+    returnType: StorageContextReadonlyValue,
+  }),
+  'Neo.StorageContext.AsReadOnly': new SimpleSysCall({
+    name: 'Neo.StorageContext.AsReadOnly',
+    args: [new SysCallArgument('context', new UnionValue([StorageContextValue, StorageContextReadonlyValue]))],
+    returnType: StorageContextReadonlyValue,
+  }),
   'Neo.Storage.Get': new SimpleSysCall({
     name: 'Neo.Storage.Get',
     args: [new SysCallArgument('context', StorageContextValue), new SysCallArgument('key', StorageKey)],
```
Comment (dicarlo2) [inline packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts]: This should take a union of `StorageContextValue` and `StorageContextReadonlyValue`

File: packages/neo-one-smart-contract/src/index.d.ts [hunk 8]
```diff
@@ -101,11 +102,13 @@ declare global {
   function syscall(name: 'Neo.Asset.GetIssuer', asset: AssetBase): Buffer;
   function syscall(name: 'Neo.Contract.GetScript', contract: ContractBase): Buffer;
   function syscall(name: 'Neo.Storage.GetContext'): StorageContextBase;
+  function syscall(name: 'Neo.Storage.GetReadOnlyContext'): StorageContextBase;
+  function syscall(name: 'Neo.StorageContext.AsReadOnly', context: StorageContextBase): StorageContextBase;
```
Comment (dicarlo2) [inline packages/neo-one-smart-contract/src/index.d.ts]: You need to regenerate this file.

File: packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts (hunk 4 above)
Comment (davemneo) [reply packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts]: I thought it was in a conversation... but can't find that -- so I'm reverting.

File: packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts [hunk 9]
```diff
@@ -260,6 +260,7 @@ const OutputValue = new BlockchainInterface('OutputBase');
 const TransactionValue = new BlockchainInterface('TransactionBase');
 const ValidatorValue = new BlockchainInterface('ValidatorBase');
 const StorageContextValue = new BlockchainInterface('StorageContextBase');
+const StorageContextReadonlyValue = new BlockchainInterface('StorageContextReadonlyBase');
```
Comment (dicarlo2) [inline packages/neo-one-smart-contract-compiler/src/compile/syscalls.ts]: nit: `StorageContextReadOnlyBase` - change `Readonly` to `ReadOnly` everywhere to be consistent with the syscall.
